packages/core/src/storages/request_queue_v2.ts

```diff
@@ -196,48 +259,54 @@ export class RequestQueue extends RequestProvider {
             return;
         }
 
-        this._listHeadAndLockPromise ??= this._listHeadAndLock().finally(() => {
-            this._listHeadAndLockPromise = null;
+        this.listHeadAndLockPromise ??= this._listHeadAndLock().finally(() => {
+            this.listHeadAndLockPromise = null;
         });
 
-        await this._listHeadAndLockPromise;
+        await this.listHeadAndLockPromise;
     }
 
     private async _listHeadAndLock(): Promise<void> {
-        const forefront = await this.hasPendingForefrontRequests();
+        const hasPendingForefrontRequests = await this.hasPendingForefrontRequests();
 
         const headData = await this.client.listAndLockHead({
-            limit: Math.min(forefront ? this.assumedForefrontCount : 25, 25),
+            limit: Math.min(hasPendingForefrontRequests ? this.assumedForefrontCount : 25, 25),
             lockSecs: this.requestLockSecs,
         });
 
+        this.queueHasLockedRequests = headData.queueHasLockedRequests;
+
         const headIdBuffer = [];
 
+        const giveUpLock = async (id?: string, uniqueKey?: string) => {
+            if (id === undefined) {
+                return;
+            }
+
+            try {
+                await this.client.deleteRequestLock(id);
+            } catch {
+                this.log.debug('Failed to delete request lock', { id, uniqueKey });
+            }
+        };
+
+        // If we tried to read new forefront requests, but another client appeared in the meantime, we can't be sure we'll only read our requests.
+        // To retain the correct queue ordering, we rollback this head read.
+        if (hasPendingForefrontRequests && headData.hadMultipleClients) {
+            this.log.debug(`Skipping this read - forefront requests may not be fully consistent`);
+            await Promise.all(headData.items.map(async ({ id, uniqueKey }) => giveUpLock(id, uniqueKey)));
+        }
```

Comment:
I was thinking about those scenarios during the last batch of forefront fixes, but I considered them too edge case-y to handle them. 

I'm not sure about these lines, though - let's say there are two clients using the queue. Won't this cause any forefront request to be locked and then immediately unlocked? `headData.hadMultipleClients` doesn't iirc say "a new client just appeared", once it's set, it's never `false` again, no? 🤔

Reply:
> I considered them too edge case-y to handle them

My point here: I'd be happy not to do this at all. Both the clients are using the same queue - as a user, I'd be fine with some request intermingling.